fdbserver/BulkDumpUtil.actor.cpp

```diff
@@ -0,0 +1,229 @@
+			ASSERT(sstWriterByteSample->finish());
+		} else {
+			ASSERT(deleteFile(byteSampleFilePath));
+		}
+	}
+
+	std::string manifestFileName = generateBulkDumpManifestFileName(dumpVersion);
+	std::string manifestFilePath = abspath(joinPath(dumpFolder, manifestFileName));
+	ASSERT(!fileExists(manifestFilePath));
+
+	BulkDumpManifest manifest(
+	    dataFilePath, manifestFilePath, byteSampleFilePath, dumpRange.begin, dumpRange.end, dumpVersion, "", dumpBytes);
+	writeFile(manifestFilePath, manifest.toString());
+
+	return manifest;
+}
+
+void bulkDumpFileCopy(std::string fromFile, std::string toFile, size_t fileBytesMax) {
+	std::string content = readFileBytes(fromFile, fileBytesMax);
+	writeFile(toFile, content);
+	return;
+}
+
+ACTOR Future<Void> bulkDumpTransportCP_impl(std::string fromRoot,
+                                            std::string toRoot,
+                                            std::string relativeFolder,
+                                            size_t fileBytesMax,
+                                            UID logId) {
+	loop {
+		state std::string toFolder = abspath(joinPath(toRoot, relativeFolder));
+		state std::string fromFolder = abspath(joinPath(fromRoot, relativeFolder));
+		state std::string fromFile;
+		state std::string toFile;
+		try {
+			// Clear existing folder
+			platform::eraseDirectoryRecursive(toFolder);
+			if (!platform::createDirectory(toFolder)) {
+				throw retry();
+			}
+
+			// Move bulk dump files to the target folder
+			for (const auto& filePath : platform::listFiles(fromFolder)) {
+				fromFile = abspath(joinPath(fromFolder, basename(filePath)));
+				toFile = abspath(joinPath(toFolder, basename(filePath)));
+				bulkDumpFileCopy(fromFile, toFile, fileBytesMax);
+				TraceEvent(SevInfo, "SSBulkDumpSSTFileCopied", logId)
+				    .detail("FromFile", fromFile)
+				    .detail("ToFile", toFile);
+			}
+			break;
+		} catch (Error& e) {
+			if (e.code() == error_code_actor_cancelled) {
+				throw e;
+			}
+			TraceEvent(SevInfo, "SSBulkDumpSSTFileCopyError", logId)
+			    .errorUnsuppressed(e)
+			    .detail("FromRoot", fromRoot)
+			    .detail("ToRoot", toRoot)
+			    .detail("RelativeFolder", relativeFolder)
+			    .detail("FromFile", fromFile)
+			    .detail("ToFile", toFile);
+			wait(delay(5.0));
+		}
+	}
+	return Void();
+}
+
+ACTOR Future<Void> uploadFiles(BulkDumpTransportMethod transportMethod,
+                               std::string fromRoot,
+                               std::string toRoot,
+                               std::string relativeFolder,
+                               UID logId) {
+	// Upload to S3 or mock file copy
+	if (transportMethod != BulkDumpTransportMethod::CP) {
+		ASSERT(false);
+		throw not_implemented();
+	}
+	wait(bulkDumpTransportCP_impl(fromRoot, toRoot, relativeFolder, SERVER_KNOBS->BULKLOAD_FILE_BYTES_MAX, logId));
+	return Void();
+}
+
+ACTOR Future<Void> persistCompleteBulkDumpRange(Database cx, BulkDumpState bulkDumpState) {
```

Comment:
Whats this method doing? It writes out a range a piece at a time?

Reply:
Yes. Mark a part of range completed. This method is called by SS.